service/src/main/java/com/google/auto/service/processor/AutoServiceProcessor.java

```diff
@@ -243,23 +253,38 @@ private String getBinaryNameImpl(TypeElement element, String className) {
     return getBinaryNameImpl(typeElement, typeElement.getSimpleName() + "$" + className);
   }
 
-  private DeclaredType getProviderInterface(AnnotationMirror providerAnnotation) {
-
-    // The very simplest of way of doing this, is also unfortunately unworkable.
-    // We'd like to do:
-    //    ServiceProvider provider = e.getAnnotation(ServiceProvider.class);
-    //    Class<?> providerInterface = provider.value();
-    //
-    // but unfortunately we can't load the arbitrary class at annotation
-    // processing time. So, instead, we have to use the mirror to get at the
-    // value (much more painful).
-
-    Map<? extends ExecutableElement, ? extends AnnotationValue> valueIndex =
-        providerAnnotation.getElementValues();
-    log("annotation values: " + valueIndex);
-
-    AnnotationValue value = valueIndex.values().iterator().next();
-    return (DeclaredType) value.getValue();
+  /**
+   * Returns the contents of a {@code Class[]}-typed field in a given {@code annotationMirror}.
+   *
+   * <p>This method is needed because directly reading the value of such a field from an
+   * AnnotationMirror throws: <pre>
+   * javax.lang.model.type.MirroredTypeException: Attempt to access Class object for TypeMirror Foo.
+   * </pre>
+   *
+   * @param annotationMirror The AnnotationMirror to read fields from, e.g. {@link AutoService}.
+   * @param fieldName The name of the field to read, e.g. "value".
+   * @return a set of fully-qualified names of classes appearing in 'fieldName' on 'annotation' on
+   *     'element'.
+   */
+  private ImmutableSet<DeclaredType> getFieldOfClasses(
+      AnnotationMirror annotationMirror,
+      String fieldName,
+      Elements elementUtils) {
+    for (Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry :
```

Comment:
I think this can be:

```java
((List<AnnovationValue>) AnnotationMirrors.getAnnotationValue(annotationMirror, "value"))
    .stream()
    .map(value -> MoreTypes.asDeclared(value.getValue())
    .collect(toImmutableSet());
```

Reply:
Done (with some extras) in dfbbdd4